### ISSUE TITLE
Featurerequest: Subnotebooks

### ISSUE THREAD

**Issue by nr458h** [laurent22/joplin#122]
### Operating system

- [ ] Windows
- [ ] macOS
- [ ] Linux
- [x] Android
- [ ] iOS

### Application

- [ ] Desktop
- [x] Mobile
- [ ] Terminal
---
To organize many notes it would be greate to be able to create subnotebooks on any level
Notebook
--Subnotebook 1
---- Subnotebook 1.1
---- Subnotebook 1.2
-------- Subnotebook 1.2.2 ....
-- Subnotebook 2

**Comment by juanreppucci** [comment]
It would be great to have hierarchical notes, I think I would extend greatly the capabilities and organization. I hope they implement this feature!!

**Comment by nr458h** [comment]
Maybe you can have a look at how it can be done with hierarchical notes in laverna

https://github.com/Laverna/laverna

**Comment by laurent22** [comment]
I'm wondering, wouldn't something like [pinned notebooks](https://github.com/laurent22/joplin/issues/296) (the equivalent of stars in Evernote) solves this issue? I think the problem is that when having many notebooks, it's hard to get to the one you need, but with pinned notebooks you could put the ones you currently need on top.

I can't imagine a workflow where one needs to browse through deep hierarchies of notebooks on a daily basis. Usually you'd be using the same few notebooks everyday, so if they could be pinned that would solve it. I could be wrong though, so feel free to comment here.

As it is, I think it's unlikely it will be implemented due to how complex it would make the UI, especially on mobile and CLI, but maybe there are other ways to make dealing with multiple notebooks more user friendly.

**Comment by RCJacH** [comment]
I use Joplin daily and would really like to have subnotebooks available for organization purposes.

The way I just it, as an example, would be:
A main notebook named Joplin for everything Joplin related,
Then subnotebooks named: Features, Code blocks, Bugs,
Then under those subnotebooks, I will have a note for each issue under the section.

This way, each note will only contain information regarding the very specific topic, while the notebook itself will not be polluted with hundreds of files regarding different aspect of Joplin. While this example may only contain one level of subnotebooks, but it might grow a lot deeper when you increase the scale of things. 

For example, you as a developer may have a lot of application projects, some already published (like Joplin) and some died in the idea stage with maybe only one line of proposition. So you may have a notebook called **Application**, then inside that a subnotebook called **Joplin**, another called **The other Notebook**, then maybe ten or twenty more applications like these, with Joplin pinned to the top of the **Application** notebook. At the same time, as Joplin gets more features, you may need subnotebooks to categorize your feature, for example **Terminal**, **UI**, **Synchronization**. 

Currently, we have all note of all aspect from all applications listed under **Application** notebook, that may contain hundreds or thousands of notes listed here (all tagged under *note* lol?). Imagine we suddenly want to look for the note that contains the idea for **The other Notebook** or any feature that had not been worked on for half a year or more (to the point that we don't even remember how it was named)? Subnotebooks will help us organize our notes in a more systematic method. We know what we are working on and we can find where a note is located right away with an ordered hierarchy.

Regarding UI, there is a Chinese notebook application called Wiz that list out folders like this:
![wiz folder](https://user-images.githubusercontent.com/12930244/37532712-c17180d2-297a-11e8-9b3a-255200a28e6e.jpg)

So basically using arrows to show the folded/unfolded state of sub-folders.

That being said, for daily uses, maybe a separated notebook called **Favourite** or **Recent**?

**Comment by nr458h** [comment]
@laurent22 I agree, that there are not that many notes you look into every day, at the moment I use a tag for that.

But for all the notes which I don't look into every day are Subnotebooks very useful, because I can group hierarchical notes, which belong together. In one single subnotebook than I can see which note I worked last time on.
Other case is, I read a lot of books and make notes to them. But to a single book especially good ones it is hard to life only with one note, or there is a task regarding what you want to do because you read this book and then without subnotebooks in my folder there would be different notes and tasks for the same book.
A third use case are context:
I do have a private life and I don't want to mix up thinks belonging work stuff or for my article so be able to folde unfolde one of this context-folders enable me to only see those notes I want to see.

But still like this up and want to say thank you for your work and making it open source!

**Comment by cb3inco** [comment]
@laurent22 I use hierarchies of notebooks in Evernotes to segment areas of life. Work, personal, etc. I would like this in Joplin if you are considering it.

**Comment by EBendinelli** [comment]
Hi there,

I've been thinking about quitting Evernote for years and I just discovered Joplin which first looked like a dream awaken. It such a great piece of software that I can't thank you enough for making it.

Just wanted to my +1 to this feature request. I was not a big fan of subnotebooks but at some point I had so many notes that I needed things to be more organized and subdivided. Having a notebook "Music production" and 6 subnotebooks for each project/part of the creative process is a must have for me.

Has anyone started working on this? I've never used Electron but I made a few things with AngularJS and some PHP frameworks, so I'd be glad to give it a try or at least help someone.

**Comment by Abijeet** [comment]
> Just wanted to my +1 to this feature request. I was not a big fan of subnotebooks but at some point I had so many notes that I needed things to be more organized and subdivided. Having a notebook "Music production" and 6 subnotebooks for each project/part of the creative process is a must have for me.

Just wondering, have you tried using tags? It works well for me.

**Comment by EBendinelli** [comment]
> Just wondering, have you tried using tags? It works well for me.

Yup, I do use tags but for a different purpose, it helps me label my notes so I can quickly find say "voice ideas" or a "synth patch" within a project (and I have those for almost all my music project). It could work to subdivide project but that would still let my notebooks pretty messy as I can have up to 100 notes for each subnotebooks in my current Evernote profile. 

I've quickly looked at the code and I feel that the best approach would be creating a new class "Pile" or "Stack" which would stand for a stack of notebooks. I think that we don't want to have notebooks which are both parents to notes and other notebooks so extending the Folder class does not seem appropriate. Creating a pile of notebook would solve the problem. What do you think?

**Comment by laurent22** [comment]
> I've quickly looked at the code and I feel that the best approach would be creating a new class "Pile" or "Stack" which would stand for a stack of notebooks. I think that we don't want to have notebooks which are both parents to notes and other notebooks so extending the Folder class does not seem appropriate. Creating a pile of notebook would solve the problem. What do you think?

The [current approach](https://github.com/laurent22/joplin/tree/subnotebooks) is to have a tree similar to a filesystem, so a notebook can have sub-notebook but also notes inside of it.

A "Stack" I guess would be a special case of this where one chooses not to have notes inside certain notebooks, but only sub-notebooks.

**Comment by laurent22** [comment]
Sub-notebooks are now supported in the desktop, mobile and CLI apps. The desktop app can both display the tree of notebooks as well as organise them via drag and drop. The mobile and terminal app currently can display the tree, but not organise the notebooks (so you would do this in the desktop app and then sync).

![](https://joplin.cozic.net/images/SubNotebooks.png)

[Check the doc](https://joplin.cozic.net/#sub-notebooks) for more information. As it's quite new there could still be some issues so in that case feel free to post here.

(Known bug: just saw the tags are displayed incorrectly on the desktop app).

**Comment by nr458h** [comment]
Many thanks for implementing this feature!! :1st_place_medal: 

I would like to add something about the UI - in my opinion it would be nice to be able to collapse or "uncollapse" not just by clicking the +/- but also by clicking the root folder-name (in your example Personal, Work,....)

**Comment by laurent22** [comment]
@nr458h, you can do so by double-clicking the folder. Single click is reserved to show the content of the folder (since it can have both notes and sub-notebooks).

**Comment by glenndebacker** [comment]
This is brilliant but one of the problems is that I cant have two subfolders with the same name e.g. I have a folder "work projects" and "personal projects" and I cant create 2x a (sub)folder "mobile apps" because it displays a message that is already exists.

**Comment by edhilgendorf** [comment]
I'm experiencing an issue where sub-notebooks do not show under the correct parent notebook in CLI. It's fine in mobile and desktop, but does not display correctly for CLI.
Example:
Two notebooks, Work and Personal, each with a couple sub-notebooks. I can create a new sub-notebook in GUI for Personal, then sync, open CLI and it shows the new sub-notebook underneath Work instead. 

Is anyone else getting a similar issue? Thanks -- I can open this as a separate issue if needed.
